File: circuit_view.py
import os
import collections
import json
import numpy
import itertools
import csv
from typing import Type, Dict, List, Any, Tuple

from comp_score_mgr import LocalScoreManager
from util import Group, Score, Stat, Rank, ScoresDict, SCORES_DIR

RAW = "raw"
NORMAL = "normal"


def build_totals(all_scores: Dict[str, Dict[str, ScoresDict]]) -> Tuple[ScoresDict, ScoresDict]:
    """
    Builds up all of the raw and normalized scores across the given competitions for all groups.
    :param all_scores: all competition scores
    :return: tuple of [raw scores dict, normalized scores dict]
    """
    all_raw: ScoresDict = {}
    all_normal: ScoresDict = {}
    for comp in all_scores:
        raw = all_scores[comp]["raw"]
        normal = all_scores[comp]["normal"]

        # TODO nicer reduce function?
        for group in raw:
            all_raw[group] = (all_raw[group] + raw[group]
                              ) if group in all_raw else raw[group]
            all_normal[group] = (
                all_normal[group] + normal[group]) if group in all_normal else normal[group]

    return all_raw, all_normal


def get_stats(scores: ScoresDict):
    """
    Converts dictionary of scores to dictionaries of median and mean values
    :param scores: dictionary of group to list of scores
    :return: median and mean dictionaries
    """
    med: Dict[Group, Stat] = {}
    mean: Dict[Group, Stat] = {}

    for group in scores:
        med[group] = Stat(numpy.median(scores[group]))
        mean[group] = Stat(numpy.mean(scores[group]))

    return med, mean


def get_ranks(stats_map: Dict[Group, Stat]) -> Dict[Group, Rank]:
    # TODO use pandas rank
    sorted_by_value = sorted(
        stats_map.items(), key=lambda kv: kv[1], reverse=True)
    # start with 1
    return {tup[0]: (i + 1) for i, tup in enumerate(sorted_by_value)}


class CircuitView:
    """
    Represents the state of the circuit.
    """

    def process(self, num: int, year: str):
        """
        Process competition scores to produce a CircuitView. `num` is the number of competitions to
        process. If num is -1, processes all competitions. `year` is the year to process.
        """
        self.year = year

        # First, convert (num, year) to comps
        with open(os.path.join(SCORES_DIR, year, "details.json")) as infile:
            _comps = json.load(infile)["order"]

        if num > len(_comps):
            raise "Illegal argument: num"

        if num < 0:
            num = len(_comps)

        self.comps = _comps[0:num]

        print("comps:")
        print(self.comps)

        self.comp_details: Dict[str, Dict[str, Any]] = {
            comp: self.handle_comp(comp) for comp in self.comps
        }

        # build normals
        raw, normal = build_totals(self.comp_details)
        self.groups = list(raw.keys())

        # evaluate numbers
        self.amed, self.amean = get_stats(raw)
        self.rmed, self.rmean = get_stats(normal)

        # get ranks
        self.amed_rank = get_ranks(self.amed)
        self.amean_rank = get_ranks(self.amean)
        self.rmed_rank = get_ranks(self.rmed)
        self.rmean_rank = get_ranks(self.rmean)

        # compute misc. stats
        self.attended: Dict[Group, List[str]] = {
            group: [
                comp for comp in self.comps if group in self.comp_details[comp][RAW]]
            for group in self.groups
        }
        self.avg_groups_per_comp = numpy.mean(
            [len(self.comp_details[comp][RAW]) for comp in self.comp_details])
        self.avg_judges_per_comp = numpy.mean(
            [len(self.comp_details[comp]["judge_avgs"]) for comp in self.comp_details])
        self.avg_comps_per_group = numpy.mean(
            [len(self.attended[group]) for group in self.groups])
        self.best_score = {
            "group": "Lel",
            "comp": "Lol",
            "score": 420.69
        }

    # TODO look up a better way to persist python classes to files
    @staticmethod
    def load(filename: str):
        with open(filename, 'r') as f:
            d = json.load(f)

        cv = CircuitView()
        cv.year = d['year']
        cv.comps = d['comps']
        cv.comp_details = d['comp_details']
        cv.groups = d['groups']
        cv.amed = d['amed']
        cv.amean = d['amean']
        cv.rmed = d['rmed']
        cv.rmean = d['rmean']
        cv.amed_rank = d['amed_rank']
        cv.amean_rank = d['amean_rank']
        cv.rmed_rank = d['rmed_rank']
        cv.rmean_rank = d['rmean_rank']
        cv.attended = d['attended']
        cv.avg_groups_per_comp = d['avg_groups_per_comp']
        cv.avg_judges_per_comp = d['avg_judges_per_comp']
        cv.avg_comps_per_group = d['avg_comps_per_group']
        cv.best_score = d['best_score']

        return cv

    def dump(self, filename: str):
        with open(filename, 'w') as f:
            json.dump(self.__dict__, f, indent=4)

    def select_groups(self, threshold: int):
        """
        Select groups given a threshold.
        """
        return [t for t in self.groups if self.amed_rank[t] <= threshold and self.amean_rank[t] <= threshold and self.rmed_rank[t] <= threshold and self.rmean_rank[t] <= threshold]

    def get_standings(self):
        """
        Returns an ordered dictionary of all of the thresholded groups.
        """
        buckets: Dict[float, List[str]] = {}
        print(f"{len(self.groups)} total groups")

        # Bucketize all groups
        for group in self.groups:
            bucket: Rank = max(self.amed_rank[group], self.amean_rank[group], self.rmed_rank[group],
                               self.rmean_rank[group])
            buckets[bucket] = (buckets[bucket] + [group]
                               ) if bucket in buckets else [group]

        # Sort each bucket by group name
        buckets = {key: sorted(value) for (key, value) in buckets.items()}

        # Convert to OrderedDict for output
        return collections.OrderedDict(sorted(iter(buckets.items())))

    def save_standings(self, ordered_buckets: collections.OrderedDict, filename: str):
        """
        Bucketize and get standings. Save to filename (CSV)
        :return:
        """
        with open(filename, mode="w") as outfile:
            outfile.write("Threshold,Groups\n")
            for bucket in ordered_buckets:
                outfile.write(
                    f"{bucket},{' '.join(ordered_buckets[bucket])}\n")

    def save_standings_json(self, ordered_buckets: collections.OrderedDict, filename: str):
        with open(filename, mode="w") as outfile:
            json.dump([
                {
                    "threshold": bucket,
                    "groups": ordered_buckets[bucket]
                }
                for bucket in ordered_buckets
            ], outfile)

    def get_group_stats(self, group: Group):
        return {
            "amed": self.amed[group] if group in self.amed else 0,
            "amean": self.amean[group] if group in self.amean else 0,
            "rmed": self.rmed[group] if group in self.rmed else 0,
            "rmean": self.rmean[group] if group in self.rmean else 0,
        }

    def get_group_ranks(self, group: Group):
        return {
            "amed": self.amed_rank[group] if group in self.amed_rank else len(self.groups) + 1,
            "amean": self.amean_rank[group] if group in self.amean_rank else len(self.groups) + 1,
            "rmed": self.rmed_rank[group] if group in self.rmed_rank else len(self.groups) + 1,
            "rmean": self.rmean_rank[group] if group in self.rmean_rank else len(self.groups) + 1,
            "total": len(self.groups),
        }

    def handle_comp(self, comp: str) -> Dict[str, Any]:
        """
        Handles a single competition.
        :param comp: name of comp
        :return: raw and normalized score dictionary, mapping group to list of scores for this comp
        """
        score_mgr = LocalScoreManager()

        raw, num_judges = score_mgr.get_raw_scores(self.year, comp)

        # normalize for each group for this comp
        judge_avgs = [numpy.mean([raw[group][i] for group in raw])
                      for i in range(num_judges)]

        normal = {group: [x * 100 / judge_avgs[i]
                          for i, x in enumerate(raw[group])] for group in raw}

        final_scores = {group: numpy.average(
            normal[group]) for group in normal}
        final_scores_list = final_scores.values()
        comp_max = max(final_scores_list)
        comp_min = min(final_scores_list)
        # TODO judge names

        return {RAW: raw, NORMAL: normal, 'final_scores': final_scores, 'max': comp_max, 'min': comp_min,
                'judge_avgs': judge_avgs}

File: app/src/circuitView.ts
import log from "loglevel";
import { median, min } from "mathjs";
import {
  reduce,
  forEach,
  concat,
  toPairs,
  reverse,
  sortBy,
  values,
  keys,
  map,
  range,
  mapValues,
  mean,
  filter,
  get,
  size,
  zipObject,
  max
} from "lodash";

import { DETAILS } from "./constants";
import { GSheetsScoreManager } from "./scoreManager";
import { ScoresDict, Group, Stat, Rank } from "./types";

/*
    Builds up all of the raw and normalized scores across the given competitions for all groups.
    :param all_scores: all competition scores
    :return: tuple of [raw scores dict, normalized scores dict]
    */
function build_totals(
  allScores?: Record<string, Record<string, ScoresDict>>
): [ScoresDict, ScoresDict] {
  const allRaw: ScoresDict = {};
  const allNormal: ScoresDict = {};

  forEach(allScores, val => {
    const raw = val.raw;
    const normal = val.normal;

    // TODO nicer reduce function?
    forEach(raw, (scores, group) => {
      allRaw[group] = group in allRaw ? concat(allRaw[group], scores) : scores;
      allNormal[group] =
        group in allNormal ? concat(allNormal[group], normal[group]) : normal[group];
    });
  });

  return [allRaw, allNormal];
}

/*
    Converts dictionary of scores to dictionaries of median and mean values
    :param scores: dictionary of group to list of scores
    :return: median and mean dictionaries
    */
function get_stats(scores?: ScoresDict) {
  const _med: Record<Group, Stat> = {};
  const _mean: Record<Group, Stat> = {};

  forEach(scores, (score, group) => {
    _med[group] = median(score);
    _mean[group] = mean(score);
  });

  return [_med, _mean];
}

// Map of group -> value
function get_ranks(statsMap: Record<Group, Stat>): Record<Group, Rank> {
  const pairs = toPairs(statsMap);
  const sortedByValue = reverse(sortBy(values(pairs), [p => p[1]]));

  // start with 1
  return reduce(
    sortedByValue,
    (acc, cur, idx) => {
      acc[cur[0]] = idx + 1;
      return acc;
    },
    {} as Record<Group, Rank>
  );
}

export class CircuitView {
  year: string;
  comps: any;
  compDetails?: Record<string, Record<string, any>>;
  groups?: string[];
  amed?: Record<string, number>;
  amean?: Record<string, number>;
  rmed?: Record<string, number>;
  rmean?: Record<string, number>;
  amedRank?: Record<string, number>;
  ameanRank?: Record<string, number>;
  rmedRank?: Record<string, number>;
  rmeanRank?: Record<string, number>;

  /* Process competition scores to produce a CircuitView. `num` is the number of competitions to process. If num is -1, processes all competitions. `year` is the year to process. */
  constructor(num: number, year: string) {
    this.year = year;

    // First, convert (num, year) to comps
    const _comps: string[] = DETAILS[year].order;

    if (num > _comps.length) {
      throw new Error("Illegal argument: num");
    }

    if (num < 0) num = _comps.length;

    this.comps = _comps.slice(0, num);

    log.debug("comps:", this.comps);
  }

  async process() {
    this.compDetails = zipObject(
      this.comps,
      await Promise.all(map(this.comps, comp => this.handle_comp(comp)))
    );

    // build normals
    const [raw, normal] = build_totals(this.compDetails);
    this.groups = keys(raw);

    // evaluate numbers
    const [amed, amean] = get_stats(raw);
    const [rmed, rmean] = get_stats(normal);
    this.amed = amed;
    this.amean = amean;
    this.rmed = rmed;
    this.rmean = rmean;

    // get ranks
    this.amedRank = get_ranks(this.amed);
    this.ameanRank = get_ranks(this.amean);
    this.rmedRank = get_ranks(this.rmed);
    this.rmeanRank = get_ranks(this.rmean);

    // compute misc. stats
    // this.attended: Record<Group, Array<string>> = reduce(this.groups, (acc, group) => {
    //     const list = [comp for comp in this.comps if group in this.comp_details[comp][RAW]];
    //     const list = filter(group in this.comp_details[comp][RAW] ?
    //     acc[group] = list
    // }, {});

    // {
    //     group: [
    //         ]
    //     for group in this.groups
    // }
    // this.avg_groups_per_comp = numpy.mean(
    //     [len(this.comp_details[comp][RAW]) for comp in this.comp_details])
    // this.avg_judges_per_comp = numpy.mean(
    //     [len(this.comp_details[comp]["judge_avgs"]) for comp in this.comp_details])
    // this.avg_comps_per_group = numpy.mean(
    //     [len(this.attended[group]) for group in this.groups])
    // this.best_score = {
    //     "group": "Lel",
    //     "comp": "Lol",
    //     "score": 420.69
    // }
  }

  /*
        Handles a single competition.
        :param comp: name of comp
        :return: raw and normalized score dictionary, mapping group to list of scores for this comp
        */
  async handle_comp(comp: string): Promise<Record<string, any>> {
    const scoreManager = new GSheetsScoreManager();

    const [raw, numJudges] = await scoreManager.get_raw_scores(this.year, comp);

    // normalize for each group for this comp
    const judgeAvgs = map(range(numJudges), i => mean(map(raw, scores => scores[i])));

    const normal = mapValues(raw, scores => map(scores, (x, i) => (x * 100) / judgeAvgs[i]));

    const finalScores = mapValues(normal, scores => mean(scores));
    const finalScoresList = values(finalScores);
    const compMax = max(finalScoresList);
    const compMin = min(finalScoresList);
    // TODO judge names

    return {
      RAW: raw,
      NORMAL: normal,
      final_scores: finalScores,
      max: compMax,
      min: compMin,
      judge_avgs: judgeAvgs
    };
  }

  /*
        Returns an ordered dictionary of all of the thresholded groups.
        */
  get_standings() {
    const buckets: Record<number, string[]> = {};
    log.debug(`${size(this.groups)} total groups`);

    // Bucketize all groups
    forEach(this.groups, group => {
      const bucket = max([
        get(this.amedRank, `[${group}]`, size(this.groups)),
        get(this.ameanRank, `[${group}]`, size(this.groups)),
        get(this.rmedRank, `[${group}]`, size(this.groups)),
        get(this.rmeanRank, `[${group}]`, size(this.groups))
      ]);
      if (bucket) {
        buckets[bucket] = bucket in buckets ? concat(buckets[bucket], group) : [group];
      }
    });

    // Sort each bucket by group name
    return mapValues(buckets, vals => sortBy(vals));
  }

  /* Select groups given a threshold. */
  selectGroups = (threshold: number) =>
    filter(
      this.groups,
      t =>
        get(this.amedRank, `[${t}]`, size(this.groups)) <= threshold &&
        get(this.ameanRank, `[${t}]`, size(this.groups)) <= threshold &&
        get(this.rmedRank, `[${t}]`, size(this.groups)) <= threshold &&
        get(this.rmeanRank, `[${t}]`, size(this.groups)) <= threshold
    );
}
